In [1]:
#hiddencell
import pandas as pd 
from pbl_tools import *

# 스테이지 2

# 1. pandas를 이용해 csv 파일 읽어오기
[문제 1]    
Pandas 라이브러리(library)를 가져와보세요.   
그리고 train.csv, test.csv, sample_submission.csv 파일을 각각 train, test, submission 변수로 읽어오세요.     
아래 빈칸을 채워주세요.

In [3]:
import ___ as ___

train = pd.___('train.csv')
test = pd.___('test.csv')
submission = pd.___('sample_submission.csv')

In [4]:
#checkcode
ensure_vals(globals(), 'train', 'test', 'submission')
@check_safety
def check(user_answer_train=train,
          user_answer_test=test, 
          user_answer_submission=submission):
    
    c_point0 = hasattr(user_answer_train, 'tail')
    c_point1 = hasattr(user_answer_test, 'tail')
    c_point2 = hasattr(user_answer_submission, 'tail')

    if (c_point0 and 
        c_point1 and 
        c_point2 
    ):
        return True
    else:
        return False

check()

True

### Inst.  

### Hint.
`pd.read_csv('파일경로/파일명')` 를 활용해 보세요.

### Solution.
```python
import pandas as pd

train = pd.read_csv('train.csv')
test = pd.read_csv('test.csv')
submission = pd.read_csv('sample_submission.csv')
```

# 2. 데이터프레임(DataFrame)정보 출력하기

In [3]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 273 entries, 0 to 272
Data columns (total 10 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   date_time                  273 non-null    object 
 1   wind_direction             273 non-null    float64
 2   sky_condition              273 non-null    float64
 3   precipitation_form         273 non-null    float64
 4   wind_speed                 273 non-null    float64
 5   humidity                   273 non-null    float64
 6   low_temp                   273 non-null    float64
 7   high_temp                  273 non-null    float64
 8   Precipitation_Probability  273 non-null    float64
 9   number_of_rentals          273 non-null    int64  
dtypes: float64(8), int64(1), object(1)
memory usage: 21.5+ KB


In [ ]:
#checkcode
#empty

### Inst.

`info()` 메소드는 pandas 라이브러리(Library)에서 제공하는 함수 중 하나로, **데이터프레임(DataFrame)의 정보를 제공** 합니다.  
데이터프레임의 행(row) 수, 칼럼(column) 수, 각 칼럼의 이름, 데이터 타입, 메모리 사용량 등을 보여주며,  
결측치(null)의 개수도 확인할 수 있습니다.  

이 코드 기법은 데이터 분석 및 전처리 과정에서 매우 유용하게 사용됩니다.  
데이터프레임의 크기와 구성을 파악하여 결측치 처리, 데이터 타입 변경, 이상치 처리 등의 작업을 수행할 수 있습니다.  
또한, 메모리 사용량을 파악하여 데이터프레임의 크기를 최적화하는데도 활용됩니다.  

`info()` 메소드의 출력 결과 중에서, 각 칼럼에 대한 정보는 다음과 같은 의미를 가집니다.  

- **column**: 데이터프레임의 **컬럼(열)** 이름을 보여줍니다.  
- **Non-null count**: 각 컬럼에 있는 **비어있지 않은 값의 개수** 를 보여줍니다.  
- **dtype**: 각 컬럼의 **데이터 타입** 을 보여줍니다. 예를 들어, object는 문자열, int는 정수, float는 소수점이 있는 숫자 등을 의미합니다.  
  
---

#### 결과 해석
train 데이터셋은 총 10개의 열(column)과 273개의 행(row)으로 이루어져 있습니다.    
또한 데이터프레임의 Non-Null Count를 살펴보면 결측치(누락된 데이터)가 존재하지 않습니다.       
이러한 정보를 바탕으로 데이터프레임의 구성과 각 열의 정보를 파악할 수 있습니다.

**[train.csv]** 
1. date_time : 일별 날짜
2. wind_direction: 풍향 (degree)
3. sky_condition : 하늘 상태
    - 맑음(1), 구름조금(2), 구름많음(3), 흐림(4)
4. precipitation_form : 강수 형태
5. wind_speed : 풍속 (m/s)
6. humidity : 습도 (%)
7. low_temp : 최저 기온 ( `C)
8. high_temp : 최고 기온 ( `C)
9. Precipitation_Probability : 강수 확률
10. **number_of_rentals : 따릉이 대여량**

### Hint.
empty

### Solution.
empty

# 3. 데이터프레임(DataFrame)의 칼럼(column) 이름 확인하기

[문제 2]  
데이터프레임(DataFrame)의 columns 속성을 이용하여 열(column)의 이름들을 반환해 봅시다.

In [4]:
train.___

Index(['date_time', 'wind_direction', 'sky_condition', 'precipitation_form',
       'wind_speed', 'humidity', 'low_temp', 'high_temp',
       'Precipitation_Probability', 'number_of_rentals'],
      dtype='object')

In [ ]:
#checkcode
ensure_vals(globals())
@check_safety
def check(user_answer=_,number=10):
    c_point0 = (len(_) == number)
    if c_point0:
        return True
    else:
        return False
check()

### Inst.
pandas 라이브러리(Library)를 사용하여 불러온 데이터셋(train)의 **열(column) 이름들을 반환** 하는 속성(attribute)입니다.  
이를 통해, 데이터셋의 열 이름들을 파악하고, 원하는 열에 쉽게 접근할 수 있습니다.  

pandas DataFrame 객체에서 제공하는 속성으로, 데이터셋의 열 이름들이 포함된 1차원 배열(Series 객체)을 반환합니다.  
이 배열에서는 각 열의 이름이 문자열로 포함되어 있으며, 각 열의 순서는 데이터프레임의 열 순서와 일치합니다.  

예를 들어, train 데이터프레임이 `'age', 'gender', 'income', 'education'` 등의 열(column)을 가지고 있다면,  
train.columns를 호출하면 Index(['age', 'gender', 'income', 'education'], dtype='object')와 같은 결과를 반환합니다.  
이를 통해, 데이터셋의 열 이름을 파악하고, 필요한 열에 쉽게 접근할 수 있습니다.  

### Hint.
empty

### Solution.
```py
train.columns
```

# 4. 데이터프레임(DataFrame)의 수치형 변수들에 대한 기술 통계량(descriptive statistics) 출력하기

[문제 3]  
데이터프레임(Dataframe)의 describe() 메서드를 이용하여 피처(feature)의 통계 값들을 가져와 출력해 봅시다.

In [5]:
train.___()

wind_direction  sky_condition  precipitation_form  wind_speed  \
count      273.000000     273.000000          273.000000  273.000000   
mean       202.750967       2.288256            0.100963    2.480963   
std         56.659232       0.961775            0.203193    0.884397   
min         57.047000       1.000000            0.000000    0.753000   
25%        171.541000       1.405000            0.000000    1.820000   
50%        209.774000       2.167000            0.000000    2.411000   
75%        238.412000       3.000000            0.088000    2.924000   
max        321.622000       4.000000            1.000000    5.607000   

         humidity    low_temp   high_temp  Precipitation_Probability  \
count  273.000000  273.000000  273.000000                 273.000000   
mean    56.745491   13.795249   23.384733                  16.878103   
std     12.351268    5.107711    5.204605                  16.643772   
min     24.831000    1.938000    9.895000                   0.000000   
25%     47.196000    9.938000   19.842000                   4.054000   
50%     55.845000   14.375000   24.158000                  12.162000   
75%     66.419000   18.000000   27.526000                  22.973000   
max     88.885000   22.312000   33.421000                  82.162000   

       number_of_rentals  
count         273.000000  
mean        59574.978022  
std         27659.575774  
min          1037.000000  
25%         36761.000000  
50%         63032.000000  
75%         81515.000000  
max        110377.000000

In [9]:
#checkcode
ensure_vals(globals())
@check_safety
def check(user_answer=_,number=(8,9)):
    c_point0 = user_answer.shape == number
    if c_point0:
        return True
    else:
        return False
check()

True

### Inst.
`describe` 메서드를 사용하면 각 피처(feature)의 기본적인 **통계 값들을 확인** 할 수 있습니다.

- **count** : 해당 피처(feature)의 데이터 **개수**
- **mean** : 해당 피처(feature)의 데이터 **평균값**
- **std** : 해당 피처(feature)의 데이터 **표준편차**
- **min** : 해당 피처(feature)의 데이터 **최솟값**
- **25%, 50%, 75%** : 해당 피처(feature)의 값들을 크기순으로 정렬했을 때, 각각 25%, 50%, 75% 지점의 백분위수
    - 백분위수(percentile)란, 크기가 있는 값들을 순서대로 나열했을 때 그 값의 상대적인 위치 를 말합니다.   
    일반적으로 크기가 작은 순으로 오름차순 정렬을 했을 때 0을 최솟값, 100을 최댓값으로 합니다.
즉, 100개의 값을 가지는 데이터가 존재할 때 25 백분위수는 그 자료의 값들 중 25번째로 작은 값을 의미합니다.
- **max** : 해당 피처(feature)의 데이터 **최댓값**


### Hint.
empty

### Solution.
```python
train.describe()
```

# 5. submission 데이터의 하위 2개의 행(row) 데이터 출력하기

[문제 4]  
pandas의 `tail()` 메소드를 이용하여 하위 2개의 행(row) 데이터를 출력해 보세요.

In [6]:
submission.___(___)

date_time  number_of_rentals
89  2021-06-29                  0
90  2021-06-30                  0

In [ ]:
#checkcode
ensure_vals(globals())
@check_safety
def check(user_answer=_,number=2):
    c_point0 = (len(user_answer) == number)
    if c_point0:
        return True
    else:
        return False
check()

### Inst.

### Hint.

`tail()` 함수는 일반적으로 데이터의 **하위 2개 행(row)** 데이터를 출력합니다.    
출력하는 행(row)의 개수를 늘리거나 줄이려면 tail(n) 함수의 인자로 원하는 개수를 입력하면 됩니다.   

### Solution.
```python
submission.tail(2)
```

# 6. target의 평균값 추출

[문제 5]  
데이터프레임(Dataframe)의 `mean()` 메서드를 이용하여 특정한 열(column)의 평균 값을 가져와서 출력해 보세요.

In [7]:
mean_target = train['number_of_rentals'].___()
mean_target

59574.97802197802

In [ ]:
#checkcode
ensure_vals(globals(),'train')
@check_safety
def check(user_answer = _):
    c_point0 = (_ == train['number_of_rentals'].mean())
    if c_point0:
        return True
    else:
        return False
check()

### Inst.

train 데이터의 target 피처(feature)에 대해 `mean()` 메서드를 적용하여 평균값을 mean_target 변수에 할당하였습니다.

### Hint.
empty

### Solution.
```python
mean_target = train['number_of_rentals'].mean()
mean_target
```

# 7. 평균값으로 submission 데이터 채우기

[문제 6]  
submission 데이터의 number_of_rentals 피처(feature)에 mean_target 값을 할당해 보세요.

In [3]:
submission['number_of_rentals'] = ___
submission.head()

date_time  number_of_rentals
0  2021-04-01       59574.978022
1  2021-04-02       59574.978022
2  2021-04-03       59574.978022
3  2021-04-04       59574.978022
4  2021-04-05       59574.978022

In [13]:
#checkcode
ensure_vals(globals(),'submission','mean_target')
@check_safety
def check(
    user_submission = submission, 
    variable = 'number_of_rentals',
    user_check = mean_target
):

    c_point0 = (user_submission[variable] == user_check).all()
    if c_point0:
        return True
    else:
        return False
check()

True

### Inst.
이번 프로젝트는 따릉이 대여량을 예측한 값으로 submission 데이터를 채우는 것이 목표라고 했었는데요.  
이번 스텝에서는 따릉이 대여량을 예측한 값으로 채우지 않고,   
train 데이터셋에서 구한 number_of_rentals의 평균값으로 채워보려고 합니다.  

train 데이터셋에서 구한 number_of_rentals의 평균값으로 채우는 이유는,  
이전 스텝에서 number_of_rentals 값이 모두 0으로 되어 있는 것을 확인했기 때문입니다.  
train 데이터셋의 number_of_rentals 열이 모두 0으로 되어 있는 경우,  
이를 그대로 제출하면 예측 성능 점수가 매우 낮게 나올 가능성이 있습니다.  

이를 방지하기 위해서는 number_of_rentals 열을 합리적으로 예측할 수 있는 다른 값을 채워서 제출해야 합니다.  
일반적으로, number_of_rentals 열의 평균값(mean)을 구해서 이를 채우는 방법이 많이 사용됩니다.  
이렇게 하면 최소한의 예측 성능은 보장됩니다.  
다만, 이를 통해 높은 예측 성능을 얻기 위해서는 전처리나 모델 선택 등의 추가적인 작업이 필요합니다.  

이를 위해서는 submission 데이터의 number_of_rentals 피처(feature)에 mean_target 값을 할당하면 됩니다.  
이때, submission['target'] = mean_target와 같이 값을 할당할 수 있습니다.  

>- submission['특성 이름'] = 대체할 값

### Hint.
empty

### Solution.
```python
submission['number_of_rentals'] = mean_target
submission.head()
```

## 8. 데이터프레임(DataFrame)을 CSV 파일로 저장하기
[문제 7]  
파일을 제출하기 위해서는 저장을 해야 합니다.  
빈칸을 채워 submission.csv 파일로 저장해 보세요.

In [ ]:
submission.___('submission.csv', index=False)

In [14]:
#checkcode
ensure_vals(globals(), 'submission')
@check_safety
def check(
    user_answer=submission
):
    answer = pd.read_csv('submission.csv')

    sub_medv = user_answer['number_of_rentals'].apply(lambda x: int(x*100))
    ans_medv = answer['number_of_rentals'].apply(lambda x: int(x*100))

    c_point0 = (sub_medv == ans_medv).all()
    if c_point0:
        return True
    else:
        return False

check()

True

### Inst.

### Hint.
submission.csv 파일로 변형한 값을 저장해 보세요.  
index 값은 원활한 제출을 위해 반드시 False로 설정해 주세요.

### Solution
```python
submission.to_csv('submission.csv', index=False)
```